File: FBControlCoreTests/Tests/Unit/FBControlCoreRunLoopTests.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>

#import "FBControlCoreValueTestCase.h"

@interface FBControlCoreRunLoopTests : FBControlCoreValueTestCase

@end

@implementation FBControlCoreRunLoopTests

- (void)testNestedAwaiting
{
  FBFuture<NSNumber *> *future = [[FBFuture
    futureWithDelay:0.1 future:[FBFuture futureWithResult:@YES]]
    onQueue:dispatch_get_main_queue() map:^(id _) {
      return [[FBFuture futureWithDelay:0.1 future:[FBFuture futureWithResult:@YES]] await:nil];
    }];

  NSError *error = nil;
  BOOL succeeded = [future await:&error] != nil;
  XCTAssertTrue(succeeded);
  XCTAssertNil(error);
}

@end

File: FBControlCoreTests/Tests/Unit/FBControlCoreRunLoopTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@testable import FBControlCore
import XCTest

final class FBControlCoreRunLoopTests: XCTestCase {
  func testNestedAwaiting() throws {
    let future = FBFuture<NSNumber>(
      delay: 0.1,
      future: FBFuture(result: NSNumber(value: true))
    )
    .onQueue(
      DispatchQueue.main,
      map: { _ -> NSNumber? in
        return try? FBFuture<NSNumber>(
          delay: 0.1,
          future: FBFuture(result: NSNumber(value: true))
        )
        .await()
      })

    let result = try future.await()
    XCTAssertNotNil(result)
  }
}
